resolver/resolver.cc

```diff
@@ -948,10 +948,75 @@ class ResolveConstantsWalk {
             transformAncestor(ctx.withOwner(klass), singleton, ancst, isInclude);
         }
 
+        // TODO: Only do this check in --stripe-mode. Currently it does the check but skips reporting the error if
+        // not in --stripe-mode.
+        //
+        // TODO: Do rigorous perf analysis.
+        if (!ctx.file.data(ctx).isPackage()) {
+            const auto ambigDef = findAnyDefinitionAmbiguousWithCurrent(ctx);
+            if (ambigDef.has_value()) {
+                if (auto e = ctx.beginError(original.loc, core::errors::Resolver::AmbiguousDefinitionError)) {
+                    e.setHeader("Class definition is ambiguous", klass.show(ctx));
+                    e.addErrorLine((*ambigDef).loc(ctx), "Previously defined here.");
+                }
+            }
+        }
+
         nesting_ = nesting_->parent;
         return tree;
     }
 
+    const std::optional<core::SymbolRef> findAnyDefinitionAmbiguousWithCurrent(core::Context ctx) {
+        if (nesting_ == nullptr) {
+            // can't be ambiguous if nesting is null
+            return std::nullopt;
+        }
+
+        auto curSym = nesting_->scope;
+        auto curNesting = nesting_->parent;
+        if (curNesting == nullptr || curNesting->scope == core::Symbols::root()) {
+            // can't be ambiguous if nested directly under root scope
+            return std::nullopt;
+        }
+
+        // find preceding symbol for current def
+        core::SymbolRef lastSym;
+        while (curSym != curNesting->scope && curSym.exists() && curSym != core::Symbols::root()) {
+            lastSym = curSym;
+            curSym = curSym.owner(ctx);
+        }
+
+        if (!curSym.exists() || (curSym.exists() && curSym == core::Symbols::root())) {
+            // preceding symbol cannot be ambiguous in these cases
+            return std::nullopt;
+        }
+        auto precedingSymForCurDef = lastSym;
+
+        if (precedingSymForCurDef.isClassOrModule() &&
+            !precedingSymForCurDef.asClassOrModuleRef().data(ctx)->isClassModuleSet()) {
```

Comment:
I really like how you wrote the first half of this function with early exits to reduce the nesting depth.  WDYT about turning this into an early exit so we can reduce the nesting depth of the latter half of the function?

Reply:
Sure!